tool-plugins/nextjs-starter/src/hooks/useAppBridge.ts

```diff
@@ -0,0 +1,277 @@
+import {
+	AppBridgeSession,
+	BeginOAuthMessagePayload,
+	CreateBeginOAuthMessagePayload,
+	CreateValidateMessagePayload,
+	PluginType,
+	PostMessageAction,
+	ValidateMessagePayload,
+} from '@/types';
+import {
+	APP_BRIDGE_ORIGIN,
+	KEY_PARENT_HOST,
+	KEY_SLUG,
+	KEY_TOKEN,
+	KEY_VALIDATED_PAYLOAD,
+} from '@/utils/const';
+import { useState, useEffect } from 'react';
+
+const getPostMessageAction = (type: PluginType): PostMessageAction => {
+	switch (type) {
+		case 'space-plugin':
+			return 'app-changed';
+		case 'tool-plugin':
+			return 'tool-changed';
+		default:
+			throw new Error(`Invalid plugin type: ${type}`);
+	}
+};
+
+const getParentHost = () => {
+	const storedHost = sessionStorage.getItem(KEY_PARENT_HOST);
+	if (storedHost) {
+		return storedHost;
+	}
+	const params = new URLSearchParams(location.search);
+	const protocol = params.get('protocol');
+	const host = params.get('host');
+	if (!protocol || !host) {
+		throw new Error('Missing `protocol` or `host` in query params');
+	}
+	return `${protocol}//${host}`;
+};
+
+const getSlug = () => {
+	const storedSlug = sessionStorage.getItem(KEY_SLUG);
+	if (storedSlug) {
+		return storedSlug;
+	}
+	const params = new URLSearchParams(location.search);
+	return params.get('slug');
+};
+
+const postMessageToParent = (payload: unknown) => {
+	window.parent.postMessage(payload, getParentHost());
+};
+
+const useAppBridgeAuth = ({
+	type,
+	authenticated,
+}: {
+	type: PluginType;
+	authenticated: () => Promise<void>;
+}) => {
+	const [status, setStatus] = useState<
+		'init' | 'authenticating' | 'authenticated' | 'error'
+	>('init');
+	const [error, setError] = useState<unknown>();
+
+	const init = async () => {
+		const isInIframe = window.top !== window.self;
+
+		if (!isInIframe) {
+			setStatus('error');
+			setError('not-in-iframe');
+			return;
+		}
+
+		if (!isAuthenticated()) {
+			sendValidateMessageToParent();
+			return;
+		}
+
+		setStatus('authenticated');
+		setError(undefined);
+
+		await authenticated();
+	};
+
+	const isAuthenticated = () => {
+		try {
+			const payload: AppBridgeSession = JSON.parse(
+				sessionStorage.getItem(KEY_VALIDATED_PAYLOAD) || '',
+			);
+			return payload && new Date().getTime() / 1000 < payload.exp;
+		} catch (err) {
+			return false;
+		}
+	};
+
+	const sendValidateMessageToParent = () => {
+		setStatus('authenticating');
+		setError(undefined);
+		const host = getParentHost();
+		const slug = getSlug();
+
+		try {
+			const payload = createValidateMessagePayload({ type, slug });
+
+			postMessageToParent(payload);
+			sessionStorage.setItem(KEY_PARENT_HOST, host);
+			sessionStorage.setItem(KEY_SLUG, slug || '');
```

Comment:
What if we verify if the slug is empty before and, in case it is, call the `setError(...)`? 🤔

Or the slug doesn't always exist? Idk 🤣

Reply:
Thanks for asking!

```
const getSlug = () => {
	const storedSlug = sessionStorage.getItem(KEY_SLUG);
	if (storedSlug) {
		return storedSlug;
	}
	const params = new URLSearchParams(location.search);
	return params.get('slug');
};
```

In practice, slug should exist, because it's given by url parameters which is done by storyfront. So I think we can trust storyfront will always provide this value. However, I've put that expression for the sake of type-check. Even if somehow it fails, with the empty slug, it will fail in the next step :)